network/p2p/scoring/subscription_provider.go

```diff
@@ -1,123 +1,152 @@
+		}
+	}
 }
 
-// getPeersByTopic returns all the peers subscribed to a topic.
-// Note that this method always returns the cached version of the subscribed peers while querying the
+// updateTopics returns all the topics in the pubsub network that this node (peer) has subscribed to.
+// Note that this method always returns the cached version of the subscribed topics while querying the
 // pubsub network for the list of topics in a goroutine. Hence, the first call to this method always returns an empty
 // list.
-// As this method is injected into GossipSub, it is vital that it never block the caller, otherwise it causes a
-// deadlock on the GossipSub.
-// Also note that, this peer itself should be subscribed to the topic, otherwise, it cannot find the list of peers
-// subscribed to the topic in the pubsub network due to an inherent limitation of GossipSub.
-func (s *SubscriptionProvider) getPeersByTopic(topic string) []peer.ID {
-	go func() {
-		// TODO: refactor this to a component manager worker once we have a startable libp2p node.
-		if _, updateInProgress := s.peersByTopicUpdating.LoadOrStore(topic, true); updateInProgress {
-			// another goroutine is already updating the list of peers for this topic
-			return
-		}
+func (s *SubscriptionProvider) updateTopics(ctx irrecoverable.SignalerContext) {
+	if updateInProgress := s.allTopicsUpdate.CompareAndSwap(false, true); updateInProgress {
+		// another goroutine is already updating the list of topics
+		return
+	}
 
-		subscribedPeers := s.tp.ListPeers(topic)
-		s.peersByTopic.Store(topic, subscribedPeers)
+	// start of critical section; protected by updateInProgress atomic flag
+	allTopics := s.topicProviderOracle().GetTopics()
+	s.logger.Trace().Msgf("all topics updated: %v", allTopics)
 
-		// remove the update flag
-		s.peersByTopicUpdating.Delete(topic)
+	// increments the update cycle of the cache; so that the previous cache entries are invalidated upon a read or write.
+	s.cache.MoveToNextUpdateCycle()
+	for _, topic := range allTopics {
+		peers := s.topicProviderOracle().ListPeers(topic)
 
-		s.logger.Trace().Str("topic", topic).Msgf("peers by topic updated: %v", subscribedPeers)
-	}()
+		for _, p := range peers {
+			if _, authorized := s.idProvider.ByPeerID(p); !authorized {
+				// peer is not authorized (staked); hence it does not have a valid role in the network; and
+				// we skip the topic update for this peer (also avoiding sybil attacks on the cache).
+				s.logger.Debug().
+					Str("remote_peer_id", p2plogging.PeerId(p)).
+					Bool(logging.KeyNetworkingSecurity, true).
+					Msg("skipping topic update for unauthorized peer")
+				continue
+			}
 
-	peerId, ok := s.peersByTopic.Load(topic)
-	if !ok {
-		return make([]peer.ID, 0)
+			updatedTopics, err := s.cache.AddTopicForPeer(p, topic)
+			if err != nil {
+				// this is an irrecoverable error; hence, we crash the node.
+				ctx.Throw(fmt.Errorf("failed to update topics for peer %s: %w", p, err))
```

Comment:
nit: rather than passing the `irrecoverable.SignalerContext`, have this return an error, and the loop can throw if an error is returned.

Reply:
https://github.com/onflow/flow-go/pull/4988/commits/a46c466ff31e2ec8be33d1d08bebc321082e07ef